% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/plotting.R
\name{plot_geneset}
\alias{plot_geneset}
\title{Plot geneset scores as cells}
\usage{
plot_geneset(
  cds,
  marker_set,
  name,
  fData_col = "gene_short_name",
  method = c("totals", "corrected"),
  reduction_method = "UMAP",
  cell_size = 0.5
)
}
\arguments{
\item{cds}{Input cell_data_set object.}

\item{marker_set}{Vector of genes in the gene_metadata DataFrame (fData) that can be found in the column 'fData_col'}

\item{name}{Name given to the geneset}

\item{fData_col}{Character string denoting the gene_metadata DataFrame (fData) column that contains the genes in marker_set1.  Default = 'gene_short_name'}

\item{cell_size}{size of point on plot}
}
\value{
Plot
}
\description{
Geneset scores are a score calculated for each single cell derived from \
more than one gene.  This function plots geneset scores using monocle3's 'plot_genes' function.

When using method 'totals', the sum of the size-factor corrected, log-normalized gene \
expression for a give set of genes is performed.  When using method 'corrected', single \
cell scores for a give gene set that have been "corrected" using 100X genes with similar \
expression levels.
}
\references{
Puram, S. V. et al. Single-Cell Transcriptomic Analysis of Primary and 
Metastatic Tumor Ecosystems in Head and Neck Cancer. Cell 171, 1611.e1â€“1611.e24 (2017).
}
